test/stateful_dataloader/test_state_dict.py

```diff
@@ -1491,5 +1491,64 @@ def test_pw(self):
         self._run(True, 4)
 
 
+class CountIterCalls(torch.utils.data.IterableDataset):
+    def __init__(self, length):
+        self.length = length
+        self.iter_calls = 0
+
+    def __iter__(self):
+        self.iter_calls += 1
+        return iter(list(range(self.length)))
+
+    def state_dict(self):
+        return {"iter_calls": self.iter_calls}
+
+    def load_state_dict(self, state_dict):
+        pass
+
+
+class TestSingleIterCalled_shard0(TestCase):
+    def _get_iter_calls(self, state):
+        if "dataset_state" in state:
+            w_states = [state]
+        else:
+            w_states = list(state["_snapshot"]["_worker_snapshots"].values())
+
+        return [x["dataset_state"]["iter_calls"] for x in w_states]
+
+    def _run_test(self, num_workers):
+        dataset = CountIterCalls(100)
+        dl = StatefulDataLoader(
+            dataset=dataset,
+            num_workers=num_workers,
+            multiprocessing_context=("forkserver" if IS_MACOS and num_workers else None),
+        )
+        iter(dl)
+        state = dl.state_dict()
+        if num_workers == 0:
+            self.assertEqual(dataset.iter_calls, 1)
+        else:
+            self.assertEqual(self._get_iter_calls(state), [1] * num_workers)
+        dl2 = StatefulDataLoader(
+            dataset=dataset,
+            num_workers=num_workers,
+            multiprocessing_context=("forkserver" if IS_MACOS and num_workers else None),
+        )
+        dl2.load_state_dict(state)
+        iter(dl2)
+        state2 = dl.state_dict()
+        if num_workers == 0:
+            self.assertEqual(dataset.iter_calls, 2)
+        else:
+            print(state2)
+            self.assertEqual(self._get_iter_calls(state2), [1] * num_workers)
```

Comment:
Hm this remains 1 in the multiprocess case because we are not getting the state0 back to main process when we load from a state?

Reply:
I suspect this is because of the change to not send worker state back after initialization

Reply:
Discussing offline, my suggestion is we always send worker state back as this is a correctness issue